From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nds-ports <nds-ports@localhost>
Date: Sun, 4 Oct 2026 12:18:33 +0200
Subject: [PATCH] Show a message when a savegame can't be written

Without an SD card nothing can be written, saving then plays the "no
way" sound and shows a message instead of crashing on a NULL file. The
slot keeps its old name and state.
---
 wl_menu.c | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/wl_menu.c b/wl_menu.c
index 9d898a1..de012eb 100644
--- a/wl_menu.c
+++ b/wl_menu.c
@@ -1527,6 +1527,8 @@ CP_SaveGame (int quick)
 
             unlink (savepath);
             file = fopen (savepath, "wb");
+            if (!file)
+                return 0;
 
             snprintf (input,sizeof(input),"%s",SaveGameNames[which]);
 
@@ -1590,9 +1592,6 @@ CP_SaveGame (int quick)
                 (LSM_X + LSItems.indent + 2, LSM_Y + which * 13 + 1, input, input, true, 31,
                  LSM_W - LSItems.indent - 30))
             {
-                SaveGamesAvail[which] = 1;
-                snprintf (SaveGameNames[which],sizeof(SaveGameNames[which]),"%s",input);
-
                 if (configdir[0])
                     snprintf (savepath, sizeof(savepath), "%s/%s", configdir, name);
                 else
@@ -1600,6 +1599,19 @@ CP_SaveGame (int quick)
 
                 unlink (savepath);
                 file = fopen (savepath, "wb");
+                if (!file)
+                {
+                    SD_PlaySound (NOWAYSND);
+                    Message ("Unable to save the game!");
+                    IN_ClearKeysDown ();
+                    IN_Ack ();
+                    DrawLoadSaveScreen (1);
+                    continue;
+                }
+
+                SaveGamesAvail[which] = 1;
+                snprintf (SaveGameNames[which],sizeof(SaveGameNames[which]),"%s",input);
+
                 fwrite (input, 32, 1, file);
                 fseek (file, 32, SEEK_SET);
 
